stdlib/seq.mli

```diff
@@ -0,0 +1,40 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*                 Simon Cruanes                                          *)
+(*                                                                        *)
+(*   Copyright 2017 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(* Module [Seq]: functional iterators *)
+
+(** @since NEXT_RELEASE *)
+
+type +'a node =
+  | Nil
+  | Cons of 'a * 'a t
+
+and 'a t = unit -> 'a node
+
+val empty : 'a t
+
+val return : 'a -> 'a t
+
+val map : ('a -> 'b) -> 'a t -> 'b t
+
+val filter : ('a -> bool) -> 'a t -> 'a t
+
+val filter_map : ('a -> 'b option) -> 'a t -> 'b t
+
+val flat_map : ('a -> 'b t) -> 'a t -> 'b t
+
+val fold_left : ('a -> 'b -> 'a) -> 'a -> 'b t -> 'a
+
+val iter : ('a -> unit) -> 'a t -> unit
```

Comment:
Fair enough.  It might be worth adding a note to seq.mli stating that the generator must be prepared to be used in a non-linear way, so that people don't expose an overly naive generator iterating over lines of a text file.